service/history/api/replication/syncworkflowstate.go

```diff
@@ -0,0 +1,298 @@
+			case *persistencepb.StateMachineTombstone_RequestCancelInitiatedEventId:
+				deletedRequestCancelInfos = append(deletedRequestCancelInfos, x.RequestCancelInitiatedEventId)
+			case *persistencepb.StateMachineTombstone_SignalExternalInitiatedEventId:
+				deletedSignalInfos = append(deletedSignalInfos, x.SignalExternalInitiatedEventId)
+			case *persistencepb.StateMachineTombstone_UpdateId:
+				deletedUpdateInfos = append(deletedUpdateInfos, x.UpdateId)
+			case *persistencepb.StateMachineTombstone_StateMachinePath:
+				deletedSubStateMachine = append(deletedSubStateMachine, x.StateMachinePath)
+			default:
+				return serviceerror.NewInvalidArgument("unknown tombstone type")
+			}
+			return nil
+		}
+		for _, tombstone := range tombstoneBatch {
+			for _, t := range tombstone.StateMachineTombstones {
+				if err := addTombstone(t); err != nil {
+					return nil, err
+				}
+			}
+		}
+		executionInfo := mu.GetExecutionInfo()
+		executionInfo.UpdateInfos = nil
+		executionInfo.SubStateMachinesByType = nil
+		mutation := &persistencepb.WorkflowMutableStateMutation{
+			UpdatedActivityInfos:       getUpdatedInfo(mu.GetPendingActivityInfos(), request.VersionedTransition),
+			UpdatedTimerInfos:          getUpdatedInfo(mu.GetPendingTimerInfos(), request.VersionedTransition),
+			UpdatedChildExecutionInfos: getUpdatedInfo(mu.GetPendingChildExecutionInfos(), request.VersionedTransition),
+			UpdatedRequestCancelInfos:  getUpdatedInfo(mu.GetPendingRequestCancelExternalInfos(), request.VersionedTransition),
+			UpdatedSignalInfos:         getUpdatedInfo(mu.GetPendingSignalExternalInfos(), request.VersionedTransition),
+			UpdatedUpdateInfos:         getUpdatedInfo(mu.GetExecutionInfo().UpdateInfos, request.VersionedTransition),
+			UpdatedSubStateMachines:    updateStateMachine,
+			DeletedActivities:          deletedActivities,
+			DeletedTimers:              deletedTimers,
+			DeletedChildExecutionInfos: deletedChildExecutionInfos,
+			DeletedRequestCancelInfos:  deletedRequestCancelInfos,
+			DeletedSignalInfos:         deletedSignalInfos,
+			DeletedUpdateInfos:         deletedUpdateInfos,
+			DeletedSubStateMachines:    deletedSubStateMachine,
+			SignalRequestedIds:         mu.GetPendingSignalRequestedIds(),
+			ExecutionInfo:              executionInfo,
+		}
+
+		events, err := getSyncStateEvents(ctx, shardContext, request.VersionHistories, mu.GetExecutionInfo().GetVersionHistories(), true)
+		if err != nil {
+			return nil, err
+		}
+		newRunInfo, err := getNewRunInfo(ctx, shardContext, workflowCache, mu)
+		if err != nil {
+			return nil, err
+		}
+		return &historyservice.SyncWorkflowStateResponse{
+			EventBatches: events,
+			Attributes: &historyservice.SyncWorkflowStateResponse_Mutation{
+				Mutation: &replicationpb.SyncWorkflowStateMutationAttributes{
+					StateMutation:                     mutation,
+					InclusiveStartVersionedTransition: request.VersionedTransition,
```

Comment:
It should be the next transition of the one in the request right.

or we can change the field to be exclusive start (since the end point in the mutable state can be regarded as inclusive) if that makes things easier. That probably also simplifies the sync logic on the target side.

Reply:
Changed to the exclusive.